File: app/src/main/java/vergecurrency/vergewallet/wallet/WalletTransactionFactory.java
package vergecurrency.vergewallet.wallet;

import vergecurrency.vergewallet.service.model.FiatRate;

public class WalletTransactionFactory {

    private float amount = 0f;
    private float fiatAmount = 0f;
    private String address = "";
    private String memo = "";
    private FiatRateTicker ticker;

    public WalletTransactionFactory(FiatRateTicker ticker) {
        this.ticker = ticker;
    }

    private void setBy(String currency, float amount) {
        if (isXVG(currency)) {
            this.amount = amount;
        } else {
            fiatAmount = amount;
        }

        update(currency);
    }

    private void update(String currency) {
        FiatRate info = ticker.getRateInfo();
        if (isXVG(currency)) {
             fiatAmount = (float) (amount * info.getPrice());
        } else {
            amount = (float) (fiatAmount / info.getPrice());
        }

    }

    private boolean isXVG(String currency) {
        return currency.equals("XVG");
    }

}

File: app/src/main/kotlin/vergecurrency/vergewallet/wallet/WalletTransactionFactory.kt
package vergecurrency.vergewallet.wallet

import vergecurrency.vergewallet.service.model.FiatRate

class WalletTransactionFactory(private val ticker: FiatRateTicker) {

    private var amount = 0f
    private var fiatAmount = 0f
    private val address = ""
    private val memo = ""

    private fun setBy(currency: String, amount: Float) {
        if (isXVG(currency)) {
            this.amount = amount
        } else {
            fiatAmount = amount
        }

        update(currency)
    }

    private fun update(currency: String) {
        val info = ticker.rateInfo
        if (isXVG(currency)) {
            fiatAmount = (amount * info!!.price).toFloat()
        } else {
            amount = (fiatAmount / info!!.price).toFloat()
        }

    }

    private fun isXVG(currency: String): Boolean {
        return currency == "XVG"
    }

}
